theano/gpuarray/basic_ops.py

```diff
@@ -1377,15 +1377,140 @@ class GpuSplit(HideC, Split):
+        codes_for_checking_outputs = []  # filled later.
+        code_for_splitting = []  # filled later.
+        code_if_sync = []  # filled later.
+        full_code_for_checking_outputs = ''  # defined later.
+        full_code_for_splitting = ''  # defined later.
+        full_code_if_sync = ''  # defined later.
+
+        main_code = """
+        ga_order x_order = (%(x)s->ga.flags & GA_C_ORDER) ? GA_C_ORDER : GA_F_ORDER;
+        int ndim = PyGpuArray_NDIM(%(x)s);
+        int axis = (int)(*(%(axis_dtype)s*)PyArray_GETPTR1(%(axis)s, 0));
+        int splits_count = PyArray_SIZE(%(splits)s);
+        size_t len_along_axis, sum_of_splits = 0, current_split_start = 0;
+        %(splits_dtype)s current_split_length = 0;
+        GpuArray view;
+        int i;
+
+        /* Check inputs. */
+
+        if (splits_count != %(expected_splits_count)s) {
+            PyErr_Format(PyExc_ValueError,
+                "GpuSplit: splits count (%%d) != expected count (%%d).", splits_count, %(expected_splits_count)s);
+            %(fail)s
+        }
+        if (axis < 0) {
+            axis += ndim;
+        }
+        if (axis < 0 || axis >= ndim) {
+            PyErr_Format(PyExc_IndexError, "GpuSplit: invalid axis %%d for a %%d-D array.", axis, ndim);
+            %(fail)s
+        }
+        len_along_axis = PyGpuArray_DIM(%(x)s, axis);
+        for (i = 0; i < splits_count; ++i) {
+            current_split_length = *(%(splits_dtype)s*)PyArray_GETPTR1(%(splits)s, i);
+            if (current_split_length < 0) {
+                PyErr_Format(PyExc_ValueError,
+                    "GpuSplit: you try to take a negative number (%%ld) of elements.", current_split_length);
+                %(fail)s
+            }
+            sum_of_splits += current_split_length;
+        }
+        if (sum_of_splits != len_along_axis) {
+            PyErr_Format(PyExc_ValueError, "GpuSplit: the splits sums to %%ld, expected %%ld.", sum_of_splits, len_along_axis);
+            %(fail)s
+        }
+
+        /* Check outputs. */
+        %(full_code_for_checking_outputs)s
+
+        /* Compute splits. */
+        if (GpuArray_view(&view, &%(x)s->ga) != GA_NO_ERROR) {
+            PyErr_SetString(PyExc_RuntimeError, "GpuSplit: unable to create a view of the input.");
+            %(fail)s
+        }
+        %(full_code_for_splitting)s
+
+        /* Free memory. */
+        GpuArray_clear(&view);
+
+        /* Code added if synchronization is enabled. */
+        %(full_code_if_sync)s
+        """
+
+        for split_index, output in enumerate(outputs):
+            # When checking output, we allocate a PyGpuArrayObject with 0 dims
+            # (that is, an object with as few memory as possible), as its GpuArray field
+            # will be cleared and re-used as a view during split operations.
+            codes_for_checking_outputs.append("""
+            if (theano_prep_output(&%(output)s, 0, NULL, %(x_typecode)s, x_order, %(x)s->context) != 0) {
```

Comment:
This function is useful if you intend on actually using the memory buffer and it can allocate new ones for you if required.  Here you just free whatever is provided as output and replace it with a view so this is wasteful.